staking/slash/double-sign.go

```diff
@@ -58,33 +55,84 @@ type Moment struct {
 	ViewID  uint64   `json:"view-id"`
 }
 
+// Copy makes a deep copy of the Moment structure
+func (m Moment) Copy() Moment {
+	cp := Moment{
+		ShardID: m.ShardID,
+		Height:  m.Height,
+		ViewID:  m.ViewID,
+	}
+	if m.Epoch != nil {
+		cp.Epoch = new(big.Int).Set(m.Epoch)
+	}
+	return cp
+}
+
 // Evidence ..
 type Evidence struct {
 	Moment
 	ConflictingVotes
 	Offender common.Address `json:"offender"`
 }
 
+// Copy makes a deep copy of the Evidence structure
+func (e Evidence) Copy() Evidence {
+	return Evidence{
+		Moment:           e.Moment.Copy(),
+		ConflictingVotes: e.ConflictingVotes.Copy(),
+		Offender:         e.Offender,
+	}
+}
+
 // ConflictingVotes ..
 type ConflictingVotes struct {
 	FirstVote  Vote `json:"first-vote"`
 	SecondVote Vote `json:"second-vote"`
 }
 
+// Copy makes a deep copy of the ConflictingVotes structure
+func (cv ConflictingVotes) Copy() ConflictingVotes {
```

Comment:
Why are all these copy functions made ?

Reply:
Simply to help with the result check in test cases.

Reply:
I agree with Edgar, to be cleaner, it's not good to add test-only code in main binary. You can implement them as a test utility instead in test code only.

Reply:
Agreed. Implemented in a module that will not be included in production code. Done.

Reply:
These Copy() code are still here?

Reply:
Thanks for reminder. Just killed them all. 👍